### PR TITLE
Fix v3 client offline support

### PR DESCRIPTION
**Pull request by alimourey**
Fixed Taiga Client v3 offline support and added a test.

### CODE REVIEW

File: taigapy/client_v3.py [hunk 1]
```diff
@@ -231,6 +231,93 @@ def _ensure_dataset_version_cached(self, permaname: str, version: int) -> bool:
...
+            id = f"{name}.{version}/{file}"
+
+        metadata = self.datafile_metadata_cache.get(id, None)
+        parsed_datafile_id = _parse_datafile_id(id)
+
+        assert (
+            parsed_datafile_id
+        ), f"{id} doesn't look like a well qualified taiga datafile ID"
+
+        canonical_id = self.canonical_id_cache.get(id, None)
```
Comment (pgm) [inline taigapy/client_v3.py:287]: This code and below looks a lot like `self._get(...)` and I think I understand why you copied it here (needed to in order to eliminate the call `self.download_to_cache()` and instead fetch the value from the cache )

At the same time, we now have two code paths depending on whether we're online or not and I worry that they might fall out of sync. And given that most of the time we'll be testing changes while online, we might notice if we make a change that breaks the offline mode. (ie: if we added another storage format)

What would you think about folding things together by adding an extra parameter to `_get()` and `download_to_cache()`? 

I'm imagining something like:

```
def def _get(self, datafile_id: str, only_use_cache=False) -> pd.DataFrame:
   ...
   path = self.download_to_cache(datafile_id, LocalFormat.HDF5_MATRIX, only_use_cache=only_use_cache)
   result = read_hdf5(path)
   ...

def download_to_cache(
        self, datafile_id: str, requested_format: Union[LocalFormat, str],
        only_use_cache = False
    ) -> str:

       canonical_id = self.get_canonical_id(datafile_id)
        key = repr((canonical_id, requested_format))
        path = self.internal_format_cache.get(key, None)
        if path:
            return path
        assert not only_use_cache, f"Expected {key} to be cached, but it was not!"
        ...
```

I think this should minimize the amount of code that is specific to when the client is offline.
     
What do you think?

File: tests/test_client_v3.py [hunk 2]
```diff
@@ -444,3 +448,90 @@ def test_upload_hdf5(
...
+                local_path=str(sample_file),
+                format=upload_format,
+                custom_metadata={},
+            )
+        ],
+    )
+    assert len(version.files) == 1
+    file = version.files[0]
+
+    def mock_is_not_connected() -> bool:
```
Comment (pgm) [inline tests/test_client_v3.py:510]: This is a great unit test. I like the use of mocking to simulate being offline.

**Review by pgm — APPROVED** [review]

File: taigapy/client_v3.py (hunk 1 above)
Comment (alimourey) [reply taigapy/client_v3.py:287]: I can definitely see how this would help keep things in sync. I'll make this change.

Comment (pgm) [reply taigapy/client_v3.py:287]: Thank you!

**Review by pgm — COMMENTED** [review]
Looks good

**Pull request merged**